### PR TITLE
refactor(launch): build + _project_spec cleanup

### PR DESCRIPTION
**Pull request by bcsherma**
Description
-----------

What does the PR do? Include a concise description of the PR contents.

This PR makes major refactors to `build.py` and `_project_spec.py` within `wandb/sdk/launch`, which encapsulate the logic for how our container image build is configured from user input and the given build context.

OH, and it also destroys `unit_tests_old/test_launch` thanks to @KyleGoyette 

- [ ] I updated CHANGELOG.md, or it's not applicable

Testing
-------
How was this PR tested?

### CODE REVIEW

File: wandb/sdk/launch/_project_spec.py [hunk 1]
```diff
@@ -134,47 +130,40 @@ def __init__(
...
 
-        override_entrypoint = overrides.get("entry_point")
-        if override_entrypoint:
-            _logger.info("Adding override entry point")
-            self.override_entrypoint = EntryPoint(
-                name=get_entrypoint_file(override_entrypoint),
-                command=override_entrypoint,
-            )
+    def init_docker(self, docker_config: Dict[str, Any]) -> None:
+        self.docker_image = docker_config.get("docker_image")
```
Comment (bcsherma) [inline wandb/sdk/launch/_project_spec.py:134]: doker image needs to be set with `or launch_spec.get("image_uri")`

File: wandb/sdk/launch/builder/context_manager.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+    --uid {uid} \
+    {user} || echo ""
+"""
+
+ENTRYPOINT_TEMPLATE = """
+ENTRYPOINT {entrypoint}
+"""
+
+
+def get_base_setup(
```
Comment (KyleGoyette) [inline wandb/sdk/launch/builder/context_manager.py:267]: With all of these additional utils in here, this feels confusing. Why move these from build into context_manager?

File: tests/pytest_tests/unit_tests/test_launch/test_launch_docker.py [hunk 3]
```diff
@@ -1,25 +0,0 @@
-import os
-from unittest.mock import MagicMock
-
-from wandb.sdk.launch._project_spec import EntryPoint
-from wandb.sdk.launch.builder.build import generate_dockerfile
-
-
-def test_buildx_not_installed(monkeypatch, tmp_path):
```
Comment (KyleGoyette) [inline tests/pytest_tests/unit_tests/test_launch/test_launch_docker.py:8]: Why remove this test? Shouldn't we test the path if buildx dne in the generated dockerfile?

File: wandb/apis/public/jobs.py [hunk 4]
```diff
@@ -123,7 +123,7 @@ def _configure_launch_project_repo(self, launch_project):
         if self._notebook_job:
             self._configure_launch_project_notebook(launch_project)
         else:
-            launch_project.set_entry_point(self._entrypoint)
+            launch_project.set_job_entry_point(self._entrypoint)
```
Comment (KyleGoyette) [inline wandb/apis/public/jobs.py:126]: Still using `set_entry_point` in image and artifact jobs

**Review by KyleGoyette — COMMENTED** [review]
Overall this feels like a step in the right direction. I don't understand why we moved all of the things we moved to context_manager it seems like the context manager is now a context manager, dockerfile generator and general utils for dockerfile generation (which is a lot like the old build file)

File: wandb/sdk/launch/builder/context_manager.py (hunk 2 above)
Comment (bcsherma) [reply wandb/sdk/launch/builder/context_manager.py]: good call I moved all of these back to build.py

File: wandb/sdk/launch/builder/context_manager.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+
+    if runner_type == "sagemaker" and not launch_project.docker_image:
+        # unless user has provided their own image, sagemaker must run as root but keep the name for workdir etc
+        return username, 0
+
+    userid = launch_project.docker_user_id or os.geteuid()
+    return username, userid
+
+
+DOCKERFILE_TEMPLATE = """
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/builder/context_manager.py]: good call, moved these under `templates/dockerfile.py`

File: wandb/sdk/launch/builder/context_manager.py [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+from wandb.sdk.launch.utils import get_current_python_version
+from wandb.util import get_module
+
+from .build import _WANDB_DOCKERFILE_NAME
+
+_logger = logging.getLogger(__name__)
+
+
+class BuildContextManager:
+    """Creates a build context for a container image from job source code."""
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/builder/context_manager.py]: well the file is always part of the build context, but I've expanded this docstring to make the role more clear. lmk what you think

File: wandb/sdk/launch/_project_spec.py [hunk 7]
```diff
@@ -434,111 +389,86 @@ def _fetch_job(self) -> None:
             raise LaunchError(
                 f"Error accessing job {self.job}: {msg} on {public_api.settings.get('base_url')}"
             )
-        job.configure_launch_project(self)
+        job.configure_launch_project(self)  # Why is this a method of the job?
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/_project_spec.py:392]: good comment- this is out of scope per our conversation.

File: tests/pytest_tests/unit_tests/test_launch/test_launch_docker.py (hunk 3 above)
Comment (bcsherma) [reply tests/pytest_tests/unit_tests/test_launch/test_launch_docker.py:8]: This has been replaced by another test in `context_manager.py`

File: wandb/sdk/launch/_project_spec.py [hunk 8]
```diff
@@ -434,111 +389,84 @@ def _fetch_job(self) -> None:
...
+            env_vars["WANDB_PROJECT"] = self.target_project
+        env_vars["WANDB_ENTITY"] = self.target_entity
+        env_vars["WANDB_LAUNCH"] = "True"
+        env_vars["WANDB_RUN_ID"] = self.run_id
+        if self.docker_image:
+            env_vars["WANDB_DOCKER"] = self.docker_image
+        if self.name is not None:
+            env_vars["WANDB_NAME"] = self.name
+        if "author" in self.launch_spec and not override_api_key:
+            env_vars["WANDB_USERNAME"] = self.launch_spec["author"]
```
Comment (KyleGoyette) [inline wandb/sdk/launch/_project_spec.py:418]: Seems like this method could use a unit test to cover all cases

File: wandb/sdk/launch/builder/build.py [hunk 9]
```diff
@@ -576,6 +51,7 @@ def join(split_command: List[str]) -> str:
     return " ".join(shlex.quote(arg.replace("'", "")) for arg in split_command)
 
 
+# Why is this in here?
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/builder/build.py:65]: I still recommend moving this out of this file

File: wandb/sdk/launch/builder/context_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/builder/context_manager.py]: looks good

File: tests/pytest_tests/unit_tests/test_launch/test_launch_docker.py (hunk 3 above)
Comment (KyleGoyette) [reply tests/pytest_tests/unit_tests/test_launch/test_launch_docker.py:8]: okay

**Review by KyleGoyette — APPROVED** [review]
LGTM an additional unit test on the envs dict would be beneficial for coverage, and there is one func that I think we both agree should be moved

**Pull request merged**